src/include/slab_containers.h

```diff
@@ -0,0 +1,727 @@
+   //
+   // Free a slot, the "freeEmpty" parameter indicates if this slab should be freed if it's emptied
+   // 
+   void freeslot(slab_t* slab, size_t idx, bool freeEmpty) {
+      //
+      // Put this slot onto the per-slab freelist
+      //
+      slab->freeMap.set(idx);
+      ++freeSlotCount;
+      this->slab_item_free(trueSlotSize);
+      if (slab->freeMap.count() == 1) {
+         //
+         // put slab onto the contianer's slab freelist
+         //
+         slab->slabHead.next = freeSlabHeads.next;
+         freeSlabHeads.next->prev = &slab->slabHead;
+         freeSlabHeads.next = &slab->slabHead;
+         slab->slabHead.prev = &freeSlabHeads;
+      }         
+      if (freeEmpty && slab->freeMap.count() == slab->size && slab != &stackSlab) {
+         //
+         // Slab is entirely free
+         //
+         slab->slabHead.next->prev = slab->slabHead.prev;
+         slab->slabHead.prev->next = slab->slabHead.next;
+         ceph_assert(freeSlotCount >= slab->size);
+         freeSlotCount -= slab->size;
+         this->slab_deallocate(slab->freeMap.count(),trueSlotSize,sizeof(slab_t),true);
+         delete[] slab;
+      }
+   }
+   //
+   // Danger, because of the my_actual_allocator hack. You can't rely on T to be correct, nor any value or offset that's
+   // derived directly or indirectly from T. We use the values saved during initialization, when T was correct.
+   //
+   void addSlab(size_t size) {
+       //
+       // I need to compute the size of this structure
+       //
+       //   struct .... {
+       //        heapSlab_t *next;
+       //        T          slots[slabSize];
+       //   };
+       //
+       // However, here sizeof(T) isn't correct [see warning above, 'T' might not be correct]
+       // so I use the sizeof(T) that's actually correct: 'trueSlotSize'
+       //
+       //
+       // Allocate the slab and free the slots within.
+       //
+       size_t total = (size * trueSlotSize) + sizeof(slab_t);
+       slab_t *slab = reinterpret_cast<slab_t *>(new char[total]);
+       initSlab(slab,size);
+   }
+   
+   slot_t *allocslot() {
+      if (freeSlabHeads.next == &freeSlabHeads) {
+         addSlab(slabSize);
+      }
+      slab_t *freeSlab = slabHeadToSlab(freeSlabHeads.next);
+      const std::size_t idx = freeSlab->freeMap.find_first_set();
+      freeSlab->freeMap.reset(idx);
+      if (freeSlab->freeMap.count() == 0) {
+         //
+         // remove slab from list
+         //
+         freeSlabHeads.next = freeSlab->slabHead.next;
+         freeSlab->slabHead.next->prev = &freeSlabHeads;
+         freeSlab->slabHead.next = nullptr;
+         freeSlab->slabHead.prev = nullptr;
+      }
+      --freeSlotCount;
+      this->slab_item_allocate(trueSlotSize);
+
+      slot_t& slot = freeSlab->slot[idx];
+      slot.slab = freeSlab;
+      return &slot;
+   }
+
+   void _reserve(size_t freeCount) {
+      if (freeSlotCount < freeCount) {
+         addSlab(freeCount - freeSlotCount);
```

Comment:
TODO: rework this in following way:
 * if necessary `reserve()` gets a single block from `malloc` and distributes it among same-sized heap slabs.
 * the size of slab can be one of two compile-time known values: `HeapSizeV` and `StackSizeV`.
 * kill `slab_t::size`.